### PR TITLE
Bigquery snippets

### PR DESCRIPTION
**Pull request by davidcavazos**
@tvalentyn Java
@aaltay Python

Added Java BigQueryIO snippets and updated the Python version to match.

### CODE REVIEW

**Comment by davidcavazos** [comment]
@melap

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    TableSchema tableSchema = new TableSchema().setFields(fields);
+    // [END BigQuerySchemaObject]
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      if (writeProject != "" && writeDataset != "" && writeTable != "") {
+        tableSpec = writeProject + ":" + writeDataset + "." + writeTable;
+      }
+
+      // [START BigQueryWrite]
```
Comment (tvalentyn) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Do we need the commented part?

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 2]
```diff
@@ -989,46 +989,93 @@ def to_entity(content):
...
-      beam.io.BigQuerySink(
-          'my-project:output.output_table',
-          schema=schema,
-          write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE,
-          create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED))
+  quotes = p | beam.Create([
+      {'source': 'Mahatma Ghandi', 'quote': 'My life is my message.'}
+  ])
+
+  quotes | beam.io.Write(beam.io.BigQuerySink(
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: AFAIK beam.io.WriteToBigQuery is a better choice than BigQuerySink and we should change examples to use that. Filed https://issues.apache.org/jira/browse/BEAM-3435 as a placeholder.

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java (hunk 1 above)
Comment (davidcavazos) [reply examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: It's not needed, but as part of a snippet, it's good for the reader to know the whole context :)

File: sdks/python/apache_beam/examples/snippets/snippets.py (hunk 2 above)
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/snippets.py]: Got it, will update. Thanks!

**Comment by davidcavazos** [comment]
Ready for review

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+/**
+ * Code snippets used in webdocs.
+ */
+public class Snippets {
+
+  @DefaultCoder(AvroCoder.class)
+  static class Quote {
+    String source;
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Can these be final?

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      // [START BigQueryReadTable]
+      PCollection<Double> maxTemperatures = p
+          .apply(BigQueryIO.readTableRows().from(tableSpec))
+          // Each row is of type TableRow
+          .apply(MapElements.into(TypeDescriptors.doubles()).via(
+              (TableRow row) -> (Double) row.get("max_temperature")));
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Nice, thanks for using Java8 here.

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+              .usingStandardSql())
+          // Each row is of type TableRow
+          .apply(MapElements.into(TypeDescriptors.doubles()).via(
+              (TableRow row) -> (Double) row.get("max_temperature")));
+      // [END BigQueryReadQueryStdSQL]
+    }
+
+    {
+      // [START BigQuerySchema]
+      String tableSchema = "source:STRING, quote:STRING";
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: I suppose this refers to usage of .withJsonSchema(String jsonSchema) - but this string is not in JSON format. What's up with this?

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    // [START BigQuerySchemaObject]
+    List<TableFieldSchema> fields = new ArrayList<>(Arrays.asList(
+        new TableFieldSchema().setName("source").setType("STRING").setMode("NULLABLE"),
+        new TableFieldSchema().setName("quote").setType("STRING").setMode("REQUIRED")));
+    TableSchema tableSchema = new TableSchema().setFields(fields);
+    // [END BigQuerySchemaObject]
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      if (writeProject != "" && writeDataset != "" && writeTable != "") {
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: What if one of them is empty but others aren't?

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+      */
+
+      PCollection<Quote> quotes = p.apply(Create.of(
+          new Quote("Mahatma Ghandi", "My life is my message.")
+      ));
+
+      quotes
+          .apply(MapElements.into(TypeDescriptor.of(TableRow.class)).via(
+              (Quote elem) -> {
+                TableRow row = new TableRow();
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: This can be abbreviated to new TableRow().set(...).set(...)

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+              .to(tableSpec)
+              .withSchema(tableSchema)
+              .withWriteDisposition(WriteDisposition.WRITE_TRUNCATE)
+              .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED));
+      // [END BigQueryWrite]
+    }
+  }
+
+  /** Helper function to format results in coGroupByKeyTuple. */
+  public static String formatCoGbkResults(
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Is this related to bigquery snippets?

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java [hunk 9]
```diff
@@ -42,7 +45,29 @@
...
+
+      BigQueryOptions options = PipelineOptionsFactory.create().as(BigQueryOptions.class);
+      options.setProject(project);
+      options.setTempLocation("gs://" + project + "/samples/temp/");
+      Pipeline p = Pipeline.create(options);
+      Snippets.modelBigQueryIO(p, project, dataset, table);
+      p.run();
+    } else {
+      Pipeline p = Pipeline.create();
+      Snippets.modelBigQueryIO(p);
```
Comment (jkff) [inline examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: p.run() ?

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java [hunk 10]
```diff
@@ -42,7 +45,29 @@
...
 
   @Rule
-  public transient TestPipeline p = TestPipeline.create();
+  public final transient TestPipeline p = TestPipeline.create();
+
+  @Test
+  public void testModelBigQueryIO() {
+    // We cannot test BigQueryIO functionality in unit tests, therefore we limit ourselves
+    // to making sure the pipeline containing BigQuery sources and sinks can be built.
+    boolean runLocally = false;
```
Comment (jkff) [inline examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: Are lines 56-65 dead code?

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java (hunk 8 above)
Comment (davidcavazos) [reply examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: No, I renamed the directory and git apparently took it as replacing it with a new file. This part is unchanged.

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java (hunk 10 above)
Comment (davidcavazos) [reply examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: Basically yes, but they are useful when testing locally. Replacing the `project`, `dataset` and `table` variables with whatever you're using allows for testing locally and making sure everything is doing what it's supposed to do

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java (hunk 9 above)
Comment (davidcavazos) [reply examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: Running this will fail due to credentials and table permissions. Therefore we limit to making sure the transforms can be built and that every validation passes. See comment at line 52.

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java (hunk 5 above)
Comment (davidcavazos) [reply examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Good catch, I copied-pasted the string format from Python. Apparently they are not compatible between Java and Python. Modifying to use actual JSON. Thanks

**Comment by davidcavazos** [comment]
Updated Java snippets to address comments, ready for review

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        }
+        public Quote(String source, String quote) {
+          this.source = source;
+          this.quote = quote;
+        }
+      }
+      */
+
+      PCollection<Quote> quotes = p.apply(Create.of(
+          new Quote("Mahatma Ghandi", "My life is my message.")
```
Comment (chamikaramj) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: s/Ghandi/Gandhi :)

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 12]
```diff
@@ -989,46 +989,93 @@ def to_entity(content):
...
+  table_schema.fields.append(source_field)
+
+  quote_field = bigquery.TableFieldSchema()
+  quote_field.name = 'quote'
+  quote_field.type = 'STRING'
+  quote_field.mode = 'REQUIRED'
+  table_schema.fields.append(quote_field)
+  # [END model_bigqueryio_schema_object]
+
+  if write_project and write_project and write_table:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: write_project and write_dataset ...

File: sdks/python/apache_beam/examples/snippets/snippets_test.py [hunk 13]
```diff
@@ -589,18 +590,30 @@ def test_model_textio_compressed(self):
...
     snippets.model_datastoreio()
 
   @unittest.skipIf(base_api is None, 'GCP dependencies are not installed')
   def test_model_bigqueryio(self):
-    # We cannot test BigQueryIO functionality in unit tests therefore we limit
+    # We cannot test BigQueryIO functionality in unit tests, therefore we limit
     # ourselves to making sure the pipeline containing BigQuery sources and
     # sinks can be built.
-    snippets.model_bigqueryio()
+    run_locally = False
```
Comment (chamikaramj) [inline sdks/python/apache_beam/examples/snippets/snippets_test.py]: Why do we have to branch below given that this doesn't change ?

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java (hunk 10 above)
Comment (chamikaramj) [inline examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: Why do we need branching below given that this is always false ?

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java (hunk 11 above)
Comment (davidcavazos) [reply examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Thanks! Changing

File: examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java (hunk 10 above)
Comment (davidcavazos) [reply examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java]: It just makes it easier for us to test locally. I'll add some instructions as comments to make it clear.

File: sdks/python/apache_beam/examples/snippets/snippets.py (hunk 12 above)
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/snippets.py]: Good catch, thanks!

File: sdks/python/apache_beam/examples/snippets/snippets_test.py (hunk 13 above)
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/snippets_test.py]: To make it easier to test locally, I added a comment clarifying this.

**Comment by davidcavazos** [comment]
Addressed review comments, ready for review. As for the "run locally" logic, let me know if you would prefer to just remove that. I left it there for ease of locally running the pipeline and making sure everything actually ran correctly since what we currently run is only to make sure the transforms can be built and all validations pass, but the pipeline itself doesn't run.

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+          .setDatasetId("samples")
+          .setTableId("weather_stations");
+      // [END BigQueryTableSpecObject]
+    }
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      // [START BigQueryReadTable]
+      PCollection<Double> maxTemperatures = p
+          .apply(BigQueryIO.readTableRows().from(tableSpec))
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Suggestion: `readTableRows()` has poor performance - consider including also a snippet of `read(SerializableFunction)` which performs much better. It's also definitely a better fit for what you're doing here (immediately extracting some information from the row). In fact I'd recommend it for *all* snippets reading from BigQuery, and maybe include just 1 snippet with readTableRows.

Comment (davidcavazos) [reply examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Got it, changing

**Comment by davidcavazos** [comment]
Updated to use `read(SerializableFunction)`. Ready for review :)

**Comment by jkff** [comment]
Java snippets LGTM; deferring to Cham on the Python snippets.

Reiterating previous question: when shall we get snippets for BigQuery writes with DynamicDestinations? In a separate PR or should that be done here?

**Comment by davidcavazos** [comment]
I can include them on this PR, I'll work on them these next days :)

**Comment by davidcavazos** [comment]
Dynamic Destinations snippet added, ready for review

File: examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+              return elem.getValue().year;
+            }
+
+            @Override
+            public TableDestination getTable(Long destination) {
+              return new TableDestination(
+                  new TableReference()
+                      .setProjectId(writeProject)
+                      .setDatasetId(writeDataset)
+                      .setTableId(writeTable + "_" + destination),
```
Comment (jkff) [inline examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: In terms of BigQuery best practices, partitioning by time is better done using partition decorators (`table$YYYYMMDD`).

**Comment by davidcavazos** [comment]
Rebased to latest version, changed `SerializableCoder.of(Double.class)` to `DoubleCoder.of()`. Python tests are broken, see [BEAM-3628](https://issues.apache.org/jira/browse/BEAM-3628). Ready for review.

**Comment by davidcavazos** [comment]
@chamikaramj does this look good to you? I think it's basically ready to merge unless there are any other comments, thanks!

**Comment by chamikaramj** [comment]
LGTM for Python. Thanks.

Merging.

**Comment by chamikaramj** [comment]
Please do a git fix-up and merge commits to single/few commit(s).

**Comment by davidcavazos** [comment]
Commits merged

**Comment by chamikaramj** [comment]
Run Java PreCommit
